src/lap_stats/lap_stats.c

```diff
@@ -82,15 +82,210 @@ static float g_distance;
+ * unconfigured which may prompt an automatic reconfiguration.
+ */
+static void reset_track()
+{
+        /*
+         * Reset items related to the track so that if the track supplied
+         * is invalid we can still do basic stuff like keep track of
+         * distance.
+         */
+        g_configured = 0;
+        g_track_status = TRACK_STATUS_WAITING_TO_CONFIG;
+        memset(&g_active_track, 0, sizeof(Track));
+
+        g_geo_circle_radius = 0;
+        g_start_finish_enabled = false;
+        g_sector_enabled = false;
+}
+
+/**
+ * The GeoTriggers used here are effectively software de-bouncing
+ * systems.  They ensure that we don't accidentally trigger an
+ * event too soon.  Think of them as a GPS based Schmitt-trigger.
+ * The tradeoff in using these is that we expect certain
+ * minimum distances be traveled by the vehicle before debouncing
+ * happens.
+ * @param track Pointer to the track that we are at.
+ * @param radius Radius of the geo circles in meters.
+ */
+static void setup_geo_triggers(const Track *track, const float radius)
+{
+        GeoPoint gp;
+        struct GeoCircle gc;
+
+        gp = getStartPoint(track);
+        gc = gc_createGeoCircle(gp, radius);
+        g_start_geo_trigger = createGeoTrigger(&gc);
+
+        /*
+         * Trip start geo trigger as unit may get powered up in start
+         * geo circle
+         */
+        geo_trigger_trip(&g_start_geo_trigger);
+
+        gp = getFinishPoint(track);
+        gc = gc_createGeoCircle(gp, radius);
+        g_finish_geo_trigger = createGeoTrigger(&gc);
+}
+
+static void update_sector_geo_circle(const int sector)
+{
+        const GeoPoint point =
+                getSectorGeoPointAtIndex(&g_active_track, sector);
+        g_geo_circles.sector = gc_createGeoCircle(point, g_geo_circle_radius);
+}
+
+/**
+ * Sets the start and finish geo circles used for start and finish lines
+ * respectively.
+ * @param track Pointer to the track where we are at.
+ * @param radius Radius of the geo circles at this track in meters.
+ */
+static void setup_geo_circles(const Track *track, const float radius)
+{
+        const GeoPoint start_p = getStartPoint(track);
+        g_geo_circles.start = gc_createGeoCircle(start_p, radius);
+
+        const GeoPoint finish_p = getFinishPoint(track);
+        g_geo_circles.finish = gc_createGeoCircle(finish_p, radius);
+
+        /* Set sector geo circle to first circle */
+        update_sector_geo_circle(0);
+}
+
+/**
+ * Tells us if we have valid start and finish points for a given track.
+ * @return true if we do, false otherwise.
+ */
+static bool isStartFinishEnabled(const Track *track)
+{
+        return isFinishPointValid(track) && isStartPointValid(track);
+}
+
+/**
+ * Tells us if we have sectors defined for this track.
+ * @return true if we do, false otherwise.
+ */
+static bool isSectorTrackingEnabled(const Track *track)
+{
+        if (!isStartFinishEnabled(track))
+                return 0;
+
+        /*
```

Comment:
ok.

On Mon, Jun 27, 2016 at 8:14 PM, Andrew Stiegmann <notifications@github.com>
wrote:

> In src/lap_stats/lap_stats.c
> <https://github.com/autosportlabs/RaceCapture-Pro_firmware/pull/705#discussion_r68692018>
> :
>
> > + */
> > +static bool isStartFinishEnabled(const Track *track)
> > +{
> > +        return isFinishPointValid(track) && isStartPointValid(track);
> > +}
> > +
> > +/**
> > + * Tells us if we have sectors defined for this track.
> > + * @return true if we do, false otherwise.
> > + */
> > +static bool isSectorTrackingEnabled(const Track *track)
> > +{
> > +        if (!isStartFinishEnabled(track))
> > +                return 0;
> > +
> > +        /*
>
> Negative. Not meant for documentation like the method comments are.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/autosportlabs/RaceCapture-Pro_firmware/pull/705/files/61367b12e274e8856b5582ce0b17db5337599935#r68692018>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADHbzVYjK53rGafFRBEHxIxRP9E08H8uks5qQJGBgaJpZM4I-v7P>
> .
>



-- 
Brent Picasso
*Autosport Labs <http://www.autosportlabs.com>*
*Technology for Race and Street*